EVS Support (Video tokens pruning)

BloodAxe opened the pull request:
## Purpose

Enable use of Efficient Video Sampling (EVS) for redundant video tokens pruning:

```python
llm = LLM(
    "nvidia/Cosmos-Reason1-7B",
    video_pruning_rate=0.75, # Prune 75% video tokens, effectively reducing TTFT 4x
    limit_mm_per_prompt={"image": 10, "video": 10},
)
```

EVS reduces TTFT and ITL by pruning less important vision tokens from the LLM:

<img width="600" height="371" alt="Time to first token (TTFT)" src="https://github.com/user-attachments/assets/181a0059-89fe-4753-8379-1cc65387f8bb" />
<img width="600" height="371" alt="Token throughput" src="https://github.com/user-attachments/assets/1ff5e349-c655-4a96-9db9-d63a4d967f8a" />


## Test Plan

- Added tests to verify inference with EVS on/off works as expected

## Test Result

## (Optional) Documentation Update

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

DarkLight1337 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1209,7 +1213,28 @@ def _gather_mm_embeddings(
-                mm_embeds.append(mm_embeds_item)
+                mm_embeds_req.append(mm_embeds_item)
+
+            if supports_multimodal_pruning(self.model) and self.model.is_multimodal_pruning_enabled():
+                should_calc_mrope_positions = True
```
Move this check outside of the loop? Then we can run the following code under `if should_calc_mrope_positions:`

BloodAxe committed:
EVS support for Qwen 2.5 VL

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge branch 'main' into feature/evs-support

BloodAxe committed:
Added missing sync op for mrope_positions buffer

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Optimize copy operation by copying only necessary part of positions

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

DarkLight1337 commented on vllm/model_executor/models/qwen2_5_vl.py:
```diff
@@ -1031,12 +1115,22 @@ def _process_image_input(
                  (merge_size * merge_size)).tolist()
 
-        return image_embeds.split(sizes)
+        image_embeds_split = image_embeds.split(sizes)
```
This method also returns inside L1106, so I think we can define a new method that wraps `_process_image_input` to apply pruning

BloodAxe replied:
Not sure I'm following you here - do you want to me to refactor the part where we take `merge_size` and `sizes` to split `image_embeds` into chunks for each media element to a standalone method? If that is the case - that is possible, although we still need these `merge_size` and `sizes` inside EVS logic, so they will be recomputed (And in the existing implementation we just reuse them)

If you suggest refactor the EVS-related method into something like `_append_rope_positions(_process_image_input(...)` and `_apply_evs_and _append_rope_positions(_process_video_input(...))` - that can be done also, although we would have to recompute a bunch of stuff inside this method and also handle it in V1 and V0 api (which IMO looks worse than existing variant):

```
    def _process_video_input_evs(self, video_embeds_split, video_input):
        grid_thw = video_input["video_grid_thw"]
        assert grid_thw.ndim == 2

        merge_size = self.visual.spatial_merge_size

        # Cast to long to match the original code
        # https://github.com/huggingface/transformers/blob/41980ce93e775f6c88500c51c8db7946fc6a2add/src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py#L491 # noqa
        second_per_grid_ts = video_input["second_per_grid_ts"].long()
        tokens_per_second = self.config.vision_config.tokens_per_second

        video_embeds_out = []
        for emb, size, video_second_per_grid_t in zip(
                video_embeds_split, grid_thw, second_per_grid_ts):
            # For each video, we compute retention mask using EVS
            retention_mask = compute_retention_mask(
                emb,
                size,
                spatial_merge_size=self.visual.spatial_merge_size,
                q=self.video_pruning_rate,
            )
            positions = compute_mrope_for_media(
                size,
                merge_size,
                tokens_per_second=tokens_per_second,
                video_second_per_grid=video_second_per_grid_t.item(),
            ).to(emb.device)

            emb = emb[retention_mask]
            positions = positions[retention_mask]
            emb = torch.cat([emb, positions], dim=1)
            video_embeds_out.append(emb)
        return tuple(video_embeds_out)
```

Then:
```
def get_multimodal_embeddings(...):
  video_embeddings = self._process_video_input(multimodal_input)
  if self.is_multimodal_pruning_enabled():
    video_embeddings = self._process_video_input_evs(video_embeddings, multimodal_input)
  ...

def get_input_embeddings_v0(...):
  ...
  if video_input is not None:
    video_embeds = self._process_video_input(video_input)
    if self.is_multimodal_pruning_enabled():
      video_embeddings = self._process_video_input_evs(video_embeddings, multimodal_input)
  ...
```

So if I were to choose - I'd leave existing version as is. But the call is yours @DarkLight1337

DarkLight1337 replied:
I mean that in the current implementation, EVS is not applied if DP is enabled. To avoid pasting the same EVS code in both DP and non-DP paths, I suggest creating a separate function that applies EVS on the output of the original function.

BloodAxe replied:
Thanks! Didn't realize there is another path for DP.
I've reworked the logic and moved EVS-related stuff into separate methods to wrap _process_video_input / _process_image_input in f440149739f462e0cadff43a6973284ea6e31bba

DarkLight1337 commented on vllm/model_executor/models/qwen2_5_vl.py:
```diff
@@ -1058,7 +1152,99 @@ def _process_video_input(
                  (merge_size * merge_size)).tolist()
 
-        return video_embeds.split(sizes)
+        video_embeds_split = video_embeds.split(sizes)
```
Same

BloodAxe committed:
Make the default come from MultiModalConfig

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe commented on vllm/engine/arg_utils.py:
```diff
@@ -435,6 +435,8 @@ class EngineArgs:
     use_tqdm_on_load: bool = LoadConfig.use_tqdm_on_load
     pt_load_map_location: str = LoadConfig.pt_load_map_location
 
+    video_pruning_rate: float = ModelConfig.video_pruning_rate
```
Addressed in 338f460d240162eb7a7e8092042ad2f73fc58f91

BloodAxe committed:
Avoid extra slicing operation when not using EVS

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1151,28 +1152,45 @@ def _execute_mm_encoder(self, scheduler_output: "SchedulerOutput"):
-            # depending on the input multimodal items.
-            curr_group_outputs = self.model.get_multimodal_embeddings(
-                **mm_kwargs_group)
+            encoder_outputs_for_modality = []
```
I've reorganized the code to avoid extra slicing when EVS is disabled just to be 100% sure we don't impact non-EVS inference 9dba0620be838b0debee503c8a0c566e2b1ce442

BloodAxe committed:
Fix support for DP in EVS

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Simplify SupportsMultiModalPruning interface:
- Remove is_multimodal_pruning_enabled() method and rely on multi_modal config.is_multimodal_pruning_enabled() instead
- Compute flag value once

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe commented on vllm/model_executor/models/interfaces.py:
```diff
@@ -120,6 +120,48 @@ def get_input_embeddings(
+    """
+    supports_multimodal_pruning: ClassVar[Literal[True]] = True
+
+    def is_multimodal_pruning_enabled(self) -> bool:
```
Fixed in d94cb7f7f0da8cf17518d4dcfccf8c46f8ef371b

DarkLight1337 reviewed (commented):
Thanks, I think the code looks good now. Regarding the micro batch size, since you still see performance benefit, I think it's ok to merge this first and optimize that later. Can you add a test to check that EVS can at least run without errors?

BloodAxe committed:
Added tests for EVS

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Added tests for EVS

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge branch 'main' into feature/evs-support

Signed-off-by: Eugene Khvedchenya <ekhvedchenya@gmail.com>

BloodAxe committed:
Added check and update mrope positions for EVS model that actually uses mrope (Future-proof support for other models that does not use MRope)

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge remote-tracking branch 'origin/feature/evs-support' into feature/evs-support-clean

DarkLight1337 commented on vllm/model_executor/models/qwen2_5_vl.py:
```diff
@@ -830,6 +837,33 @@ def get_hf_processor(self, **kwargs: object) -> Qwen2_5_VLProcessor:
+            num_frames=num_frames,
+            image_processor=image_processor,
+        )
+        # Intentionally commented - although it's tempting to write:
+        # mm_config = self.ctx.model_config.get_multimodal_config()
+        # video_pruning_rate = mm_config.video_pruning_rate
+        # if video_pruning_rate is not None and video_pruning_rate > 0:
+        #     retain_faction = 1 - video_pruning_rate
+        #     num_video_tokens = int(num_video_tokens * retain_faction)
+        # This logic would not be handled correctly by a scheduler.
+        # The maximum number of video frames and encoder budged should be
+        # deducted from the _input_ media size (and not the number of output
+        # tokens as it is now)
```
Can you comment out the whole method instead?

BloodAxe replied:
I've removed unnecessary comments in a34818d24d0eac437a353fdc757625596944adc2

As for the Qwen2-VL support - I'm not against in in general, yet to be 100% sure the implementation is correct I would have to re-run the benchmarks to ensure I did everything right (both for 2 & 2.5 models) which is somewhat time consiming. 

Would it be a deal breaker to accept this MR as is and add EVS support in future? 
We are going to add EVS support for https://github.com/vllm-project/vllm/pull/23753/files, and other models in future.

BloodAxe committed:
Remove method with unnecessary comments

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge branch 'main' into feature/evs-support

ywang96 committed:
Merge branch 'main' into feature/evs-support

BloodAxe committed:
Added NVidia copyright

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge remote-tracking branch 'origin/feature/evs-support' into feature/evs-support-clean

BloodAxe committed:
Merge branch 'refs/heads/main' into feature/evs-support-clean

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

# Conflicts:
#	vllm/model_executor/models/qwen2_5_vl.py

BloodAxe committed:
Merge branch 'main' into feature/evs-support

BloodAxe committed:
Revert vllm/_custom_ops.py

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge remote-tracking branch 'origin/feature/evs-support' into feature/evs-support-clean

BloodAxe committed:
Fix EVS logic of computing number of retained tokens:
1) Make sure we always include first video frame regardless of the pruning rate
2) Use the stable sort when picking tokens to retain

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Update with SPDX

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Bugfix: Wrong clamping

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge branch 'refs/heads/main' into feature/evs-support-clean

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

# Conflicts:
#	vllm/config/__init__.py

BloodAxe committed:
Update initializaiton of video_pruning_rate

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Ensure we cast input video size to int to avoid inconsistencies in rounding between torch.long and int types.

Sometimes method has input as tuple of ints, sometimes - as torch.LongTensor and results may vary because of rounding behavior in torch-vs-python.

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge branch 'main' into feature/evs-support

BloodAxe commented on vllm/config/__init__.py:
Fixed

BloodAxe committed:
Merge branch 'main' into feature/evs-support-clean

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

# Conflicts:
#	vllm/config/__init__.py
#	vllm/model_executor/models/qwen2_5_vl.py
#	vllm/v1/worker/gpu_model_runner.py

BloodAxe committed:
Post-merge fixes for video_pruning_rate argument

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Post-merge fixes for video_pruning_rate argument

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Post-merge fixes for video_pruning_rate argument

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge branch 'main' into feature/evs-support

BloodAxe committed:
Update docstrings to make CI happy

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge remote-tracking branch 'origin/feature/evs-support' into feature/evs-support-clean

BloodAxe committed:
Update docstrings to make CI happy

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge branch 'main' into feature/evs-support

BloodAxe committed:
Merge branch 'main' into feature/evs-support-clean

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

# Conflicts:
#	vllm/v1/worker/gpu_model_runner.py

BloodAxe committed:
Update docstrings to make CI happy

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge branch 'main' into feature/evs-support

BloodAxe committed:
Update docstrings to make CI happy

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge remote-tracking branch 'origin/feature/evs-support' into feature/evs-support-clean

BloodAxe committed:
Fix intialization of is_multimodal_pruning_enabled

Signed-off-by: Eugene Khvedchenia <ekhvedchenia@nvidia.com>

BloodAxe committed:
Merge branch 'main' into feature/evs-support

DarkLight1337 reviewed (approved):
Sorry for the delay, let's get this merged
